include/nbl/builtin/hlsl/type_traits.hlsl

```diff
@@ -4,13 +4,425 @@
 #ifndef _NBL_BUILTIN_HLSL_TYPE_TRAITS_INCLUDED_
 #define _NBL_BUILTIN_HLSL_TYPE_TRAITS_INCLUDED_
 
+// Since HLSL currently doesnt allow type aliases we declare them as seperate structs thus they are (WORKAROUND)s
+/*
+  // helper class
+  template<class T, T v> struct integral_constant; (DONE)
+ 
+  template<bool B>
+    using bool_constant = integral_constant<bool, B>; (WORKDAROUND)
+  using true_type  = bool_constant<true>; (WORKDAROUND)
+  using false_type = bool_constant<false>; (WORKDAROUND)
+ 
+  // primary type categories
+  template<class T> struct is_void; (DONE)
+  template<class T> struct is_null_pointer; (NOT-APPLICABLE)
+  template<class T> struct is_integral; (DONE)
+  template<class T> struct is_floating_point;  (DONE)
+  template<class T> struct is_array; (DONE)
+  template<class T> struct is_pointer; (TODO)
+  template<class T> struct is_lvalue_reference; (TODO)
+  template<class T> struct is_rvalue_reference; (TODO)
+  template<class T> struct is_member_object_pointer; (TODO)
+  template<class T> struct is_member_function_pointer; (TODO)
+  template<class T> struct is_enum;  (NOT-APPLICABLE)
+  template<class T> struct is_union; (NOT-APPLICABLE)
+  template<class T> struct is_class; (NOT-APPLICABLE)
+  template<class T> struct is_function; (NOT-APPLICABLE)
+ 
+  // composite type categories
+  template<class T> struct is_reference; (TODO)
+  template<class T> struct is_arithmetic; (DONE)
+  template<class T> struct is_fundamental; (TODO)
+  template<class T> struct is_object; (NOT-APPLICABLE)
+  template<class T> struct is_scalar; (DONE)
+  template<class T> struct is_compound; (DONE)
+  template<class T> struct is_member_pointer; (TODO)
+ 
+  // type properties
+  template<class T> struct is_const; (DONE)
+  template<class T> struct is_volatile; (DONE)
+  template<class T> struct is_trivial; (EVERYTHING IS)
+  template<class T> struct is_trivially_copyable; (EVERYTHING IS)
+  template<class T> struct is_standard_layout; (NOT-APPLICABLE)
+  template<class T> struct is_empty; (DONE? sizeof(T) == 0)
+  template<class T> struct is_polymorphic; (NOTHING IS)
+  template<class T> struct is_abstract; (NOTHING IS)
+  template<class T> struct is_final; (NOTHING IS)
+  template<class T> struct is_aggregate; (DONE)
+ 
+  template<class T> struct is_signed; (DONE)
+  template<class T> struct is_unsigned; (DONE)
+  template<class T> struct is_bounded_array;
+  template<class T> struct is_unbounded_array;
+  template<class T> struct is_scoped_enum; (NOT-APPLICABLE)
+
+  // type property queries
+  template<class T> struct alignment_of; (TODO)
+  template<class T> struct rank; (TODO)
+  template<class T, unsigned I = 0> struct extent; (TODO)
```

Comment:
I want these, but in a separate PR, some time later, either implement or open an issue